game/procurement.py

```diff
@@ -116,53 +127,278 @@ def repair_runways(self, budget: float) -> float:
         return budget
 
     def random_affordable_ground_unit(
-        self, budget: float, cp: ControlPoint
+        self, budget: int, vehicle_type
     ) -> Optional[Type[VehicleType]]:
         affordable_units = [
             u
             for u in self.faction.frontline_units + self.faction.artillery_units
             if db.PRICES[u] <= budget
         ]
 
-        total_number_aa = (
-            cp.base.total_frontline_aa + cp.pending_frontline_aa_deliveries_count
-        )
-        total_non_aa = (
-            cp.base.total_armor + cp.pending_deliveries_count - total_number_aa
-        )
-        max_aa = math.ceil(total_non_aa / 8)
-
-        # Limit the number of AA units the AI will buy
-        if not total_number_aa < max_aa:
-            for unit in [u for u in affordable_units if u in TYPE_SHORAD]:
-                affordable_units.remove(unit)
+        for unit in [u for u in affordable_units if u not in vehicle_type]:
+            affordable_units.remove(unit)
 
         if not affordable_units:
             return None
         return random.choice(affordable_units)
 
-    def reinforce_front_line(self, budget: float) -> float:
+    def reinforce_front_line(self, ground_unit_budget: float) -> float:
         if not self.faction.frontline_units and not self.faction.artillery_units:
-            return budget
+            return ground_unit_budget
 
         # TODO: Attempt to transfer from reserves.
 
-        while budget > 0:
-            candidates = self.front_line_candidates()
```

Comment:
I think we do need to keep the loop. The budget should not necessarily be evenly split, and we shouldn't necessarily spend the whole budget. If there are two destinations here, one with 0 units and one with 49 units, the second only needs to buy one unit and should then stop buying for that CP, and the remaining budget for that CP should be spent on the under garrisoned CP.

If you go that route, I think you can simplify some more as well. You won't need to know exact purchase ratios, all you need to know is whether a specific unit type is over or under provisioned. `calculate_vehicle_investment_ratio` can become:


```
# GroundUnitClass is an enum you should create as well.
@staticmethod
def under_provisioned_unit_type(cp: ControlPoint) -> GroundUnitClass:
    for unit_class in GroundUnitClass:
        # You'll need to add units_of_class to Base. Could add `cp.base.unit_class_ratio(...)`
        # as well if you prefer.
        current_ratio = cp.base.total_units_of_class(unit_class) / cp.base.total_armor
        desired_ratio = self.faction.doctrine.ground_unit_procurement_ratios[unit_class]
        if current_ratio < desired_ratio:
            return unit_class
    # The forces are balanced but not at the cap. Might as well buy a tank.
    return GroundUnitClass.Tank

def random_affordable_ground_unit(self, budget: float, unit_class: GroundUnitClass) -> Optional[Type[VehicleType]]:
    ...
```

Anything I'm missing? It seems like a simpler approach to me, but I haven't fully absorbed this so I may have overlooked something. If so, lmk. If not, I'm holding off on reviewing the rest in depth since I suspect you'll be able to delete a ton of it :)

Reply:
I uploaded my next improvement: I kept the precise calculation since I want to buy not the first underrepresented unit, but the one thats most needed. Therefore the calculations must be extensive.

But I changed the logic so, that the cp with the least units gets to buy a unit first. Also I moved the limit to the settings class since I dont like magic numbers that much and assume that it is a performance thing.

Reply:
I will try it out when I get some free time.
The difference I see at the moment would be, that I calculate the ratio based on unit costs, and you make it by unit count. Which is, of course easier, but I like the ratio based on cost better since based on cost is considering differences in timeperiods where some unit types are way cheapter. But thats just my opinion.

Reply:
Pushed a commit for the refactor. I think it both behaves the way you intended as is the simpler implementation I wanted? PTAL